adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1226,4 +1300,110 @@ private static MqttConnectReturnCode getConnectReturnCode(final Throwable e) {
             return MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED;
         }
     }
+
+
+    /**
+     * A class that contains handlers for command subscription, unsubscription and sending of commands.
+     */
+    protected class CommandHandler {
+
+        private final Map<String, TriTuple<CommandSubscription, MessageConsumer, Object>> subscriptions = new ConcurrentHashMap<>();
+        private final Map<Integer, TriTuple<Long, CommandSubscription, CommandContext>> waitingForAcknowledgement = new ConcurrentHashMap<>();
+
+        /**
+         * Invoked when a device sends an MQTT <em>PUBACK</em> packet.
+         *
+         * @param msgId The msgId of the command published with QoS 1.
+         */
+        protected void handlePubAck(final Integer msgId) {
+            LOG.trace("Acknowledgement received for command [Msg-id: {}] that has been sent to device.", msgId);
+            Optional.ofNullable(removeFromWaitingForAcknowledgement(msgId)).ifPresent(value -> {
+                cancelTimer(value.one());
+
+                final CommandSubscription subscription = value.two();
+                final CommandContext commandContext = value.three();
+                afterCommandPublished(subscription, commandContext);
```

Comment:
this class is still tightly coupled to its surrounding class by invoking this method, isn't it?